lib/libc/sys/ioctl.c

```diff
@@ -88,9 +92,22 @@ _ioctl(int fd, unsigned long com, ...)
 		} else {
 			data = va_arg(ap, void *);
 		}
-		va_end(ap);
 	} else {
-		data = NULL;
+#ifdef __CHERI_PURE_CAPABILITY__
+		if (cheri_getlen((void*)ap) == sizeof(void *))
+			data = va_arg(ap, void *);
+		else if (cheri_getlen((void*)ap) == sizeof(uint64_t))
+			data = (void *)(intptr_t)va_arg(ap,  int64_t);
```

Comment:
Probably best to 's/uint64_t/int64_t' on the live above for consistency.  There is also an extra space here after 'ap'.